Add support for all parenthesized binary additions of interpolated strings

333fred opened the pull request:
Implements the decision from https://github.com/dotnet/csharplang/issues/5106. Test plan at https://github.com/dotnet/roslyn/issues/51499.

333fred commented:
@jcouv @dotnet/roslyn-compiler for reviews.

333fred commented:
@jcouv @dotnet/roslyn-compiler for review please.

jcouv commented on src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:
```diff
@@ -12794,12 +12794,100 @@ public void ParenthesizedAdditiveExpression_02()
+  IL_0041:  constrained. ""CustomHandler""
+  IL_0047:  callvirt   ""string object.ToString()""
+  IL_004c:  call       ""void System.Console.WriteLine(string)""
+  IL_0051:  ret
```
The IL is identical to `ParenthesizedAdditiveExpression_01` above. Consider merging tests.
Are there situations where the parens will affect the IL? If so, that would be an interesting test.

333fred replied:
> Are there situations where the parens will affect the IL? If so, that would be an interesting test.

If there are, that is a bug :)

jcouv commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodeExtensions.cs:
```diff
@@ -94,5 +97,163 @@ private class ContainsAwaitVisitor : BoundTreeWalkerWithStackGuardWithoutRecursi
             }
         }
+
+        public static bool VisitBinaryOperatorInterpolatedString<TArg, TInterpolatedStringType>(
```
nit: It may be good to document order of visit. For the leaves, we call `visitor` from left to right. For the binary nodes, we call `binaryOperatorCallback in depth-first postorder if I understood correctly?
I think a recursive implementation would have been easier to follow, but I assume you're worried about recursion depth.

AlekseyTs replied:
>TArg [](http://example.com/codeflow?start=65&length=4)

Usually ```TArg``` is the last type parameter

AlekseyTs replied:
>public static bool VisitBinaryOperatorInterpolatedString<TArg, TInterpolatedStringType>( [](http://example.com/codeflow?start=8&length=88)

Please provide clear documentation what call backs are supposed to be invoked and in what relative order.

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -408,32 +372,13 @@ static BoundInterpolatedString createInterpolation(BoundExpression expression, I
-            BoundBinaryOperator? current = binaryOperator;
-
-            while (current != null)
+            binaryOperator.VisitBinaryOperatorInterpolatedString(partsArrayBuilder, (BoundUnconvertedInterpolatedString unconvertedInterpolatedString, ArrayBuilder<ImmutableArray<BoundExpression>> partsArrayBuilder) =>
```
>, [](http://example.com/codeflow?start=82&length=1)

static for the `visitor` lambda? (comment applies to other callers of `VisitBinaryOperatorInterpolatedString` too)

AlekseyTs replied:
>(BoundUnconvertedInterpolatedString unconvertedInterpolatedString, ArrayBuilder<ImmutableArray<BoundExpression>> partsArrayBuilder) => [](http://example.com/codeflow?start=84&length=134)

Consider aligning the header of the lambda with its body

jcouv commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -1381,6 +1351,57 @@ IBinaryOperation createBoundUserDefinedConditionalLogicalOperator(BoundUserDefin
+            return boundBinaryOperator.RewriteInterpolatedStringAddition((this, boundBinaryOperator.InterpolatedStringHandlerData.GetValueOrDefault()), createInterpolatedString, createBinaryOperator);
+
+            static IInterpolatedStringOperation createInterpolatedStringOperand(
+                (CSharpOperationFactory @this, InterpolatedStringHandlerData Data) arg,
```
>@this [](http://example.com/codeflow?start=40&length=5)

nit: could be `This` given naming convention on tuple elements

333fred replied:
It could be, but I prefer `@this`.

jcouv commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodeExtensions.cs:
```diff
@@ -94,5 +97,163 @@ private class ContainsAwaitVisitor : BoundTreeWalkerWithStackGuardWithoutRecursi
+            Func<TArg, BoundBinaryOperator, TResult, TResult, TResult> binaryOperatorFactory)
+            where TInterpolatedStringType : BoundExpression
+        {
+            Debug.Assert(typeof(TInterpolatedStringType) == typeof(BoundUnconvertedInterpolatedString) || typeof(TInterpolatedStringType) == typeof(BoundInterpolatedString));
```
Might it be possible to remove the `TInterpolatedStringType` type parameter and use `BoundInterpolatedStringBase` instead? It's abstract and only has those two derived types. 
(same question applies to `VisitBinaryOperatorInterpolatedString`)

333fred replied:
I'll tighten the constraint, but I'd prefer to keep the type parameter to ensure that we're being specific about the allowed things and fail when we encounter something we didn't expect.

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -306,47 +272,45 @@ private bool TryBindUnconvertedBinaryOperatorToDefaultInterpolatedStringHandler(
                     original.HasErrors);
 
-            static BoundInterpolatedString createInterpolation(BoundExpression expression, ImmutableArray<BoundExpression> parts)
+            static BoundInterpolatedString createInterpolation((ImmutableArray<ImmutableArray<BoundExpression>> AppendCalls, TypeSymbol _) arg, BoundExpression expression, int i)
```
The type of the `expression` parameter could be made `BoundUnconvertedInterpolatedString` (narrower).
nit: for readability, consider moving this local function above `createBinaryOperator` (ie. swap the two local functions) to match usage order in the call to `RewriteInterpolatedStringAddition`
nit: consider moving `arg` to be the last parameter. That's more typical. Something like `Func<TInterpolatedStringType, int, TArg, TResult> interpolatedStringFactory` and `Func<BoundBinaryOperator, TResult, TResult, TArg, TResult> binaryOperatorFactory` in `RewriteInterpolatedStringAddition`

jcouv commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodeExtensions.cs:
```diff
@@ -94,5 +97,163 @@ private class ContainsAwaitVisitor : BoundTreeWalkerWithStackGuardWithoutRecursi
+            Debug.Assert(typeof(TInterpolatedStringType) == typeof(BoundUnconvertedInterpolatedString) || typeof(TInterpolatedStringType) == typeof(BoundInterpolatedString));
+            var originalStack = ArrayBuilder<BoundBinaryOperator>.GetInstance();
+            var rewrittenLefts = ArrayBuilder<(BoundExpression original, TResult rewritten)>.GetInstance();
+            (BoundBinaryOperator? original, TResult? rewritten) result = default;
```
Feels like a tuple is unnecessary here (doesn't help readability)

jcouv commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodeExtensions.cs:
```diff
@@ -94,5 +97,163 @@ private class ContainsAwaitVisitor : BoundTreeWalkerWithStackGuardWithoutRecursi
+        /// Rewriter for the BoundBinaryOperator parts fo the binary operator. Passed the callback parameter, the original binary operator, and
+        /// the rewritten left and right components.
+        /// </param>
+        public static TResult RewriteInterpolatedStringAddition<TArg, TInterpolatedStringType, TResult>(
```
Would it be simpler if we leveraged `VisitBinaryOperatorInterpolatedString` and avoided a separate implementation of traversal? 
If I understood the traversal order of `VisitBinaryOperatorInterpolatedString` correctly, then the following should work: 
  When we encounter a leaf, we rewrite it and push the rewritten node to a stack. 
  When we encounter a binary node, we pop two rewritten nodes from the stack, rewrite the binary node with those, and push the rewritten binary node.
  After the traversal is complete, the stack should contain exactly one entry for the rewrittern result.

Having a visit helper with a single callback for all nodes (as suggested above) might make that logic even clearer in terms of ordering.

(I didn't review fully this method yet)

333fred replied:
We talked offline. The Visit version needs to be depth-first in-order, so it can't be used here. Instead, I'm going to switch this to only use iteration for the left side of the tree, and recurse on the right side when necessary. Since we are generally only concerned with large trees on the left side, this should be no worse than other sections of our code for stack behavior.

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -306,47 +272,45 @@ private bool TryBindUnconvertedBinaryOperatorToDefaultInterpolatedStringHandler(
 
-            var bottomOperator = stack.Pop();
-            var result = createBinaryOperator(bottomOperator, createInterpolation(bottomOperator.Left, appendCalls[0]), rightIndex: 1);
+            Func<(ImmutableArray<ImmutableArray<BoundExpression>>, TypeSymbol), BoundUnconvertedInterpolatedString, int, BoundExpression> interpolationFactory =
```
Are these locals useful? (here and other similar call sites)

333fred replied:
The types of the local functions are more strict than the delegate types themselves (BoundBinaryOperator vs BoundExpression) so I'd need to either cast or include the type parameters. This seemed the cleanest way to me.

jcouv replied:
I didn't follow that. Feels like the local functions should just use the same signature.

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -230,71 +231,36 @@ private bool TryBindUnconvertedBinaryOperatorToDefaultInterpolatedStringHandler(
-                var rightInterpolatedString = (BoundUnconvertedInterpolatedString)current.Right;
-
-                if (!ValidateInterpolatedStringParts(rightInterpolatedString))
+            if (!binaryOperator.VisitBinaryOperatorInterpolatedString(partsArrayBuilder, static (BoundUnconvertedInterpolatedString unconvertedInterpolatedString, ArrayBuilder<ImmutableArray<BoundExpression>> partsArrayBuilder) =>
```
>static [](http://example.com/codeflow?start=89&length=6)

nit: consider using a local function (`validateAndPushParts` or some such)

AlekseyTs replied:
>static (BoundUnconvertedInterpolatedString unconvertedInterpolatedString, ArrayBuilder<ImmutableArray<BoundExpression>> partsArrayBuilder) => [](http://example.com/codeflow?start=89&length=141)

Consider placing the lambda header on a separate line. This should make it easier to figure out what the block on the next line is about. Alternatively the body can be aligned with the header.

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -230,71 +231,36 @@ private bool TryBindUnconvertedBinaryOperatorToDefaultInterpolatedStringHandler(
 
             if (isConstant ||
-                (partsCount <= 4 && partsArrayBuilder.All(static parts => AllInterpolatedStringPartsAreStrings(parts))))
+                (partsArrayBuilder.Count <= 4 && partsArrayBuilder.All(static parts => AllInterpolatedStringPartsAreStrings(parts))))
```
>AllInterpolatedStringPartsAreStrings [](http://example.com/codeflow?start=87&length=36)

Does creating a lambda rather than just referencing `AllInterpolatedStringPartsAreStrings` avoid delegate type instantiations here?

333fred replied:
Yes, we never cache method references.

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs:
```diff
@@ -154,33 +154,12 @@ private static ImmutableArray<BoundExpression> CollectBinaryOperatorInterpolated
             Debug.Assert(node.InterpolatedStringHandlerData is not null);
             var partsBuilder = ArrayBuilder<BoundExpression>.GetInstance();
-            while (true)
+            node.VisitBinaryOperatorInterpolatedString(partsBuilder, (BoundInterpolatedString interpolatedString, ArrayBuilder<BoundExpression> partsBuilder) =>
```
I'm really liking the new helper methods to abstract the traversal. It much simplifies the different places that need to visit. Thanks for making that change

jcouv replied:
static lambda here too?

jcouv commented on src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:
```diff
@@ -12711,15 +12711,17 @@ public void InterpolatedStringsAddedUnderObjectAddition_DefiniteAssignment()
-        [Fact]
-        public void ParenthesizedAdditiveExpression_01()
+        [Theory]
+        [InlineData(@"($""{i1}"" + $""{i2}"") + $""{i3}""")]
```
Consider adding more complicated/deep additive expressions to test diverse tree shapes.
Also consider testing one very deep additive expression (that benefits from non-recursive design).

jcouv commented on src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:
```diff
@@ -2519,33 +2523,20 @@ protected void VisitBinaryInterpolatedStringAddition(BoundBinaryOperator node)
-                    throw ExceptionUtilities.UnexpectedValue(node.Left.Kind);
-            }
-        }
+        protected virtual void OnPushBinaryOperatorInterpolatedStringChildren(BoundBinaryOperator node) { }
```
Could we assert that we're dealing with an additive interpolated string expression?

333fred replied:
Not easily from this location. There's no flag that indicates that at all levels of the tree, only the top binary operation that has the InterpolatedStringHandlerData could assert this.

jcouv reviewed (commented):
Done with review pass (iteration 4)

AlekseyTs commented:
>             // check for cases 1 and 2: if they are present, we let normal string binary operator binding machinery handle it. Otherwise, we take care of it ourselves.

It is not clear what cases does this comment refer to.

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:218 in f089c9a. [](commit_id = f089c9ade9f7a47269b8e98dce776b42e29cc0eb, deletion_comment = False)

AlekseyTs commented:
>             if (InExpressionTree)

I do not see speclet ([https://github.com/dotnet/csharplang/blob/main/proposals/csharp-10.0/improved-interpolated-strings.md](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-10.0/improved-interpolated-strings.md)) mentioning expression trees. What am I missing?

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:222 in f089c9a. [](commit_id = f089c9ade9f7a47269b8e98dce776b42e29cc0eb, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -230,71 +231,36 @@ private bool TryBindUnconvertedBinaryOperatorToDefaultInterpolatedStringHandler(
-            var stack = ArrayBuilder<BoundBinaryOperator>.GetInstance();
+            // The constant value is folded as part of creating the unconverted operator. If there is a constant value, then the top-level binary operator
+            // will have one.
+            bool isConstant = binaryOperator.ConstantValue is not null;
```
>bool isConstant = binaryOperator.ConstantValue is not null; [](http://example.com/codeflow?start=12&length=59)

Consider moving the declaration to the line right above the first use of the local.

333fred replied:
I'll actually move the use here, so we don't do unnecessary walks when there's a constant value.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -230,71 +231,36 @@ private bool TryBindUnconvertedBinaryOperatorToDefaultInterpolatedStringHandler(
-                            partsArrayBuilder.Free();
-                            return false;
-                        }
+                    if (!ValidateInterpolatedStringParts(unconvertedInterpolatedString))
```
>ValidateInterpolatedStringParts [](http://example.com/codeflow?start=25&length=31)

The name of the helper feels confusing, unless parts flagged by it always cause an error.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -408,32 +372,13 @@ static BoundInterpolatedString createInterpolation(BoundExpression expression, I
-
-            Debug.Assert(partsArrayBuilder.Count == stack.Count + 1);
-            Debug.Assert(partsArrayBuilder.Count >= 2);
+                partsArrayBuilder.Add(unconvertedInterpolatedString.Parts);
```
>unconvertedInterpolatedString.Parts [](http://example.com/codeflow?start=38&length=35)

Would be useful to assert something about the parts? Validity, etc.

333fred replied:
I can't think of anything useful to assert here. The parts can have errors or be invalid in the context (such as having `awaits` within a `ref struct` builder type) and binding will issue errors as appropriate.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:
```diff
@@ -727,7 +727,12 @@ private BoundExpression RebindSimpleBinaryOperatorAsConverted(BoundBinaryOperato
             while (stack.TryPop(out current))
             {
-                Debug.Assert(current.Right is BoundUnconvertedInterpolatedString);
+                var right = current.Right switch
```
>var right = current.Right switch [](http://example.com/codeflow?start=16&length=32)

The local isn't used. However, it looks like the intent is to  use it in place of ```current.Right``` in the next statement. Given that the tests are passing, I assume we are missing tests for "interesting" scenarios.

333fred replied:
It's actually not necessary at the moment, but more through what I would call a fluke of how the binder works than through intentional design. `BindSimpleBinaryOperator` will convert both the left and the right to the operand types of the operator (`string` in this case), which will end up back in this path on the right side, ensuring that no `BoundUnconvertedInterpolatedString`s are in the final tree. With the way that the code is currently implemented, I don't believe there is a way to actually see an observable difference between doing the work here and letting a recursive call eventually make its way back through this path, but as I said that feels more like unintended consequence than intentional behavior, so I will fix this bug. I'll also add a test that would break, if the behavior weren't already eventually consistent.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -230,71 +231,36 @@ private bool TryBindUnconvertedBinaryOperatorToDefaultInterpolatedStringHandler(
-            partsArrayBuilder.ReverseContents();
-
             // Case 3. Bind as handler.
             var (appendCalls, data) = BindUnconvertedInterpolatedPartsToHandlerType(
```
>BindUnconvertedInterpolatedPartsToHandlerType [](http://example.com/codeflow?start=38&length=45)

Even if all the previous checks pass, is it possible that this operation fails and reports errorrs? Constructor is missing, specific Appens overload is missing, etc.

333fred replied:
This is possible, but only if the user is defining their own DefaultInterpolatedStringHandler type. They'd also see the same errors if they used the string on its own, not in a binary operator scenario.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:
```diff
@@ -727,7 +727,12 @@ private BoundExpression RebindSimpleBinaryOperatorAsConverted(BoundBinaryOperato
+                var right = current.Right switch
+                {
+                    BoundUnconvertedInterpolatedString s => s,
+                    BoundBinaryOperator b => RebindSimpleBinaryOperatorAsConverted(b, diagnostics),
```
>RebindSimpleBinaryOperatorAsConverted(b, diagnostics) [](http://example.com/codeflow?start=45&length=53)

It looks like the algorithm is not symmetrical. It applies this transformation on the right, but never on the left.

333fred replied:
I'm not sure what you mean. This is the recursive call for the right side.

AlekseyTs replied:
> I'm not sure what you mean. This is the recursive call for the right side.

The recursive call will ```TryBindUnconvertedBinaryOperatorToDefaultInterpolatedStringHandler```. So subexpressions on the right might potentially use default handler, but similar sub expressions on the left might wouldn't.

333fred replied:
Ah, I see what you're saying there. Yes, that's a good catch (and I will add a test that would bind differently in that case).

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodeExtensions.cs:
```diff
@@ -94,5 +97,163 @@ private class ContainsAwaitVisitor : BoundTreeWalkerWithStackGuardWithoutRecursi
+                switch (current.Left)
+                {
+                    case BoundBinaryOperator op:
+                        binaryOperatorCallback?.Invoke(op, arg);
```
>binaryOperatorCallback [](http://example.com/codeflow?start=24&length=22)

It looks like the ```binaryOperatorCallback``` is never called for an input that has TInterpolatedStringType on both sides. Is this correct? If so, is this intended and why?

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodeExtensions.cs:
```diff
@@ -94,5 +97,163 @@ private class ContainsAwaitVisitor : BoundTreeWalkerWithStackGuardWithoutRecursi
+                switch (current.Right)
+                {
+                    case BoundBinaryOperator rightOperator:
+                        binaryOperatorCallback?.Invoke(rightOperator, arg);
```
>binaryOperatorCallback?.Invoke(rightOperator, arg); [](http://example.com/codeflow?start=24&length=51)

It looks like the algorith is not symmetrical. For some binary operators it invokes callback after invoking visitor  and the callbacks for children, but for others it invokes the callback before.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodeExtensions.cs:
```diff
@@ -94,5 +97,163 @@ private class ContainsAwaitVisitor : BoundTreeWalkerWithStackGuardWithoutRecursi
+
+            pushLeftNodes(binary, stack);
+
+            while (stack.TryPop(out BoundBinaryOperator? current))
```
>while (stack.TryPop(out BoundBinaryOperator? current)) [](http://example.com/codeflow?start=12&length=54)

It looks like there are several "holes" in the implementation. I recommend trying to rewrite it using the following strategy: start with recursive version, then switch to manual stack tracking.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodeExtensions.cs:
```diff
@@ -94,5 +97,163 @@ private class ContainsAwaitVisitor : BoundTreeWalkerWithStackGuardWithoutRecursi
+            pushLeftNodes(binary, originalStack);
+
+            int i = 0;
+            while (originalStack.TryPeek(out var currentBinary))
```
>while (originalStack.TryPeek(out var currentBinary)) [](http://example.com/codeflow?start=12&length=52)

It feels like the implementation (if correct) is too complicated and therefore the correctness is hard to prove. Let's discuss offline ways to simplify the implementation.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:
```diff
@@ -2519,33 +2523,20 @@ protected void VisitBinaryInterpolatedStringAddition(BoundBinaryOperator node)
-                    throw ExceptionUtilities.UnexpectedValue(node.Left.Kind);
-            }
-        }
+        protected virtual void OnPushBinaryOperatorInterpolatedStringChildren(BoundBinaryOperator node) { }
```
>OnPushBinaryOperatorInterpolatedStringChildren [](http://example.com/codeflow?start=31&length=46)

Given that the default implementation is empty, I have no idea what is the purpose of this method and why it has this name.

333fred replied:
I'll rename the method to be more clear.

AlekseyTs commented:
Done with review pass (commit 4)

333fred commented:
>             // check for cases 1 and 2: if they are present, we let normal string binary operator binding machinery handle it. Otherwise, we take care of it ourselves.

It refers to the comment in the method it mentioned above, covering the different cases of interpolation binding.

---
In reply to: [920376697](https://github.com/dotnet/roslyn/pull/56333#issuecomment-920376697) [](http://example.com/codeflow?ancestors=920376697,920376697)

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:218 in f089c9a. [](commit_id = f089c9ade9f7a47269b8e98dce776b42e29cc0eb, deletion_comment = False)

333fred commented:
>             if (InExpressionTree)

`ref struct`s are not permitted in expression trees, and the default handler is a ref struct. So we need a different lowering strategy for that case.

---
In reply to: [920380977](https://github.com/dotnet/roslyn/pull/56333#issuecomment-920380977) [](http://example.com/codeflow?ancestors=920380977,920380977)

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:222 in f089c9a. [](commit_id = f089c9ade9f7a47269b8e98dce776b42e29cc0eb, deletion_comment = False)

jcouv commented on src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:
```diff
@@ -2502,15 +2502,19 @@ bool learnFromOperator(BoundBinaryOperator binary)
-            }
+            node.VisitBinaryOperatorInterpolatedString(
+                (parts, @this: this),
+                stringCallback: (BoundInterpolatedString interpolatedString, (ArrayBuilder<BoundInterpolatedString> parts, AbstractFlowPass<TLocalState, TLocalFunctionState> @this) arg) =>
```
static lambdas here too

333fred commented:
@jcouv @AlekseyTs addressed your feedback. For `RewriteInterpolatedStringAddition`, I'd suggest not diff reviewing the change and just look at the new implementation, the diff is incomprehensible.

jcouv commented on src/Compilers/Core/Portable/Collections/ArrayBuilderExtensions.cs:
```diff
@@ -218,6 +218,18 @@ public static T Peek<T>(this ArrayBuilder<T> builder)
             return builder[builder.Count - 1];
         }
 
+        public static bool TryPeek<T>(this ArrayBuilder<T> builder, [MaybeNullWhen(false)] out T result)
```
Is this new extension used?

jcouv commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodeExtensions.cs:
```diff
@@ -94,5 +97,153 @@ private class ContainsAwaitVisitor : BoundTreeWalkerWithStackGuardWithoutRecursi
+                    var rewrittenRight = currentBinary.Right switch
+                    {
+                        TInterpolatedStringType interpolatedString => interpolatedStringFactory(interpolatedString, i++, arg),
+                        BoundBinaryOperator binaryOperator => doRewrite(binaryOperator, arg, interpolatedStringFactory, binaryOperatorFactory, originalStack, ref i),
```
Re-using the same `originalStack` here feels dangerous/wrong, as `doRewrite` eventually returns the rewritten version of the last node on the stack, but in this recursive case that's not the one we want.

jcouv reviewed (commented):
Done with review pass (iteration 6)

jcouv reviewed (approved):
LGTM Thanks (iteration 7) with suggestion to make a couple more lambdas static

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:
```diff
@@ -715,26 +715,37 @@ private BoundExpression RebindSimpleBinaryOperatorAsConverted(BoundBinaryOperato
+                    var right = current.Right switch
+                    {
+                        BoundUnconvertedInterpolatedString s => s,
+                        BoundBinaryOperator b => doRebind(diagnostics, stack, b),
```
>doRebind [](http://example.com/codeflow?start=49&length=8)

It looks like this call is not going to return until it purges the entire stack, which contains entries for parent nodes. When that happens the call doesn't return the node that corresponds to the ```current.Right```, but a node that corresponds to its parent, or quite possibly some node that doesn't correspond to any original node. It doesn't feel like desired behavior.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodeExtensions.cs:
```diff
@@ -94,5 +97,152 @@ private class ContainsAwaitVisitor : BoundTreeWalkerWithStackGuardWithoutRecursi
+                        throw ExceptionUtilities.UnexpectedValue(current.Left.Kind);
+                }
+
+                binaryOperatorCallback?.Invoke(current, arg);
```
>binaryOperatorCallback?.Invoke(current, arg); [](http://example.com/codeflow?start=16&length=45)

Invoking a call back for the parent before visting right child doesn't quite match what I expect from a depth-first visit.

333fred replied:
Left, middle, right is what defines a depth-first in-order visit, which is what is documented for this function to perform.

AlekseyTs replied:
Is this order chosen just to avoid a recursion on the right? In that case, I would not optimize for that scenario and stick to the order (left, right, middle) as, I believe,  we usually do. We usually do not visit parent node in between visting different children.

333fred replied:
This order is chosen to ensure that speculation in the nullable walker will follow a lexical ordering. IE, if you put your position in the middle of a `+`, the speculative model will give you the correct result of the left having been evaluated, but not the right.

AlekseyTs replied:
> if you put your position in the middle of a +, the speculative model will give you the correct result of the left having been evaluated, but not the right.

Could you please elaborate how exactly is this achieved?

333fred replied:
I went to explain how it would be achieved, then realized that isn't how it would work, because I was misremembering where the `+` syntax start is. I'll switch to a pre-order visit.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodeExtensions.cs:
```diff
@@ -94,5 +97,152 @@ private class ContainsAwaitVisitor : BoundTreeWalkerWithStackGuardWithoutRecursi
         }
+
+        /// <summary>
+        /// Visits the binary operator tree of interpolated string additions in a depth-first in-order visit.
```
>depth-first in-order [](http://example.com/codeflow?start=82&length=20)

I think it would be clearer to specify the order explicitly. There are three nodes (left, right, parent), in what order are they supposed to be visited?

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodeExtensions.cs:
```diff
@@ -94,5 +97,152 @@ private class ContainsAwaitVisitor : BoundTreeWalkerWithStackGuardWithoutRecursi
+
+        /// <summary>
+        /// Rewrites a BoundBinaryOperator composed of interpolated strings (either converted or unconverted) iteratively, without
+        /// recursion on the left side of the tree. Nodes of the tree are rewritten in a depth-first post-order fashion.
```
>depth-first post-order [](http://example.com/codeflow?start=89&length=22)

I think it would be clearer to specify the order explicitly. There are three nodes (left, right, parent), in what order are they supposed to be visited?

AlekseyTs commented:
Done with review pass (commit 7)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [923453621](https://github.com/dotnet/roslyn/pull/56333#issuecomment-923453621) [](http://example.com/codeflow?ancestors=923453621,923453621)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:
```diff
@@ -712,29 +712,38 @@ private BoundExpression RebindSimpleBinaryOperatorAsConverted(BoundBinaryOperato
-                stack.Push(current);
-                current = current.Left as BoundBinaryOperator;
-            }
+                var stack = ArrayBuilder<BoundBinaryOperator>.GetInstance();
```
>var stack = ArrayBuilder<BoundBinaryOperator>.GetInstance(); [](http://example.com/codeflow?start=16&length=60)

I expected to see a unit-test that backs the fix.

AlekseyTs reviewed (approved):
LGTM (commit 8), modulo the missing unit-test.
